lib/collaborative_filtering.py

```diff
@@ -2,3 +2,212 @@
 """
 Module that provides the main functionalities of collaborative filtering.
 """
+
+from util.data_parser import DataParser
+from lib.evaluator import Evaluator
+from lib.abstract_recommender import AbstractRecommender
+from numpy.linalg import solve
+import numpy
+import sys
+import os
+sys.path.insert(1, os.path.join(sys.path[0], '..'))
+
+
+class CollaborativeFiltering(AbstractRecommender):
+    """
+    A class that takes in the rating matrix and outputs user and item
+    representation in latent space.
+    """
+    def __init__(self, ratings, evaluator, config, verbose=False):
+        """
+        Train a matrix factorization model to predict empty
+        entries in a matrix. The terminology assumes a
+        ratings matrix which is ~ user x item
+        @param (ndarray) ratings a matrix containign the ratings
+        1 indicates user has the document in his library
+        0 indicates otherwise.
+        @param (dict) config hyperparameters of the recommender, contains
+        _lambda and n_factors
+        @param (object) evaluator object that evaluates the recommender
+        @param (bool) verbose if True, tracing will be printed
+        """
+        self.ratings = ratings
+        self.n_users, self.n_items = ratings.shape
+        self.n_factors = config['n_factors']
```

Comment:
should you remove these 2 lines from here and use set_config? I was suggesting to remove set_config from the abstract class, as usually it will be a part of initialization.